pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -521,7 +538,35 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(logger klog.Logger, pInfo *
 	for plugin := range pInfo.UnschedulablePlugins {
 		metrics.UnschedulableReason(plugin, pInfo.Pod.Spec.SchedulerName).Inc()
 	}
-	if p.moveRequestCycle >= podSchedulingCycle {
+
+	// Check if some events, which may make this in flight Pod schedulable, happen during in-flight.
+	var moveRequestCycle int64
+	// AddUnschedulableIfNotPresent is called with the Pod at the end of scheduling or binding.
+	// So, given pInfo should have been Pop()ed before,
+	// we can assume pInfo must be recorded in inFlightPods.
+	if inFlightPod, ok := p.inFlightPods[pInfo.Pod.UID]; ok {
+		for _, e := range inFlightPod.events {
+			if len(pInfo.UnschedulablePlugins) != 0 && !p.podMatchesEvent(pInfo, e.event) {
+				// This event won't make this Pod schedulable.
+				continue
+			}
+			moveRequestCycle = e.requestCycle
+			if len(pInfo.UnschedulablePlugins) == 0 {
+				// use the latest requestCycle when no unschedulable plugin is recorded.
+				break
+			}
+		}
+	} else {
+		// It shouldn't reach here unless there is a bug somewhere.
+		// But, set podSchedulingCycle to moveRequestCycle
+		// so that this Pod won't stuck in the unschedulable pod pool.
+		logger.Error(nil, "In flight Pod isn't found in the scheduling queue. If you see this error log, it's likely a bug in the scheduler.", "pod", klog.KObj(pod))
+		moveRequestCycle = podSchedulingCycle
+	}
+
+	// TODO(sanposhiho): filter out events when the callback function, which filters out useless events, is implemented in EventToRegister.
+
+	if moveRequestCycle >= podSchedulingCycle {
```

Comment:
Should `AddUnschedulableIfNotPresent` call Done internally?

It might prevent some race conditions.

Reply:
OK, I'll change Done() so that
- [if Pod scheduling/binding is failed at some point] call Done() in `AddUnschedulableIfNotPresent`
- [if Pod is scheduled + bound successfully] call Done() in schedule_one.go